### ISSUE TITLE
Exception messages in Python not showing nicely

### ISSUE THREAD

**Issue by deaktomi** [Azure/azure-sdk-for-python#21760]
I generated a client with Autorest for C#, Node.js, and for Python as well. When my API is returning an error, I send back an503 status code with some error details in a JSON.

When I use my Node.js library and meet an error, I have a nice error message on the output of my script. See the second line
```
C:\Program Files\nodejs\node.exe .\index.js
(node:29056) UnhandledPromiseRejectionWarning: Error: {"code":400,"error":{"global":["MY-ERROR-MESSAGE-THAT-IS-COMING-FROM-THE-API"]}}
    at new RestError (c:\Users\Jorgos\Downloads\MyTests\node_modules\@azure\ms-rest-js\dist\msRest.node.js:2480:28)
    at c:\Users\Jorgos\Downloads\MyTests\node_modules\@azure\ms-rest-js\dist\msRest.node.js:3687:37
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async c:\Users\Jorgos\Downloads\MyTests\index.js:16:13
(Use `node --trace-warnings ...` to show where the warning was created)
(node:29056) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:29056) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

When I use the Python client, the error message is not so nice
```Traceback (most recent call last):
  File "C:/Users/Jorgos/Dev/Workspace/my-test/python/main.py", line 16, in <module>
    client.my_method(body=MyMethodRequest(my_parameter="myValue"))
  File "c:\users\jorgos\appdata\roaming\python\python38\site-packages\my-company\my-package\my_client.py", line 1137, in my_method
    raise models.ProblemResponseException(self._deserialize, response)
my-company.my-client.models.problem_response_py3.ProblemResponseException: Operation returned an invalid status code 'Bad Request'

Process finished with exit code 1
```

The issue is that the users of the library don't know what is their issue when they use Python. There is only a status code, but there is no error message printed, such as in the case with the Node.js package

I try to prevent putting the example scripts in a try-catch block and printing the error message in the catch block. Because I would like to keep the example scripts as clean as possible.

Do you have any suggestions on what to do here? Or do I miss something?

**Comment by brnbs** [comment]
+1 we need this feature ASAP

**Comment by techwithtim** [comment]
+1

**Comment by iscai-msft** [comment]
@brnbs @techwithtim I'm starting work on a PR now!

**Comment by deaktomi** [comment]
Thank you so much! Does it mean that it can be released in December?

**Comment by iscai-msft** [comment]
hey @deaktomi we have a PR open for it now, will give more updates on release later. Thanks so much for your patience!

**Comment by iscai-msft** [comment]
@deaktomi @brnbs @techwithtim thank you so much for your patience over the holidays! @xiangyan99 and I just released azure-core [1.21.0](https://pypi.org/project/azure-core/) on pypi with the new exception messages! Please give it a try and let us know if there are any issues!

**Comment by iscai-msft** [comment]
hey @deaktomi @brnbs @techwithtim we received some feedback and we need to revisit the way we're exposing the response body. I'm really sorry for the back and forth, reopening this issue to acknowledge that we haven't fully met your ask yet. In the meantime, a documented workaround that will get users the response body is:

```python
from azure.core.exceptions import HttpResponseError
try:
    client.make_call()
except HttpResponseError as e:
    print("\n".join([str(error), err.response.text()]))
```

**Comment by deaktomi** [comment]
Do you have any ETA on it when it will be fixed?

**Comment by iscai-msft** [comment]
hey @deaktomi sorry for the delay in response, we were off for break. We decided to go back on our release that added the error body to the response message, and have been having continued discussions about how to expose error bodies in messages. We're very sorry for this back and forth, and there is a current workaround that you can document for your users to use: see the above comment.

We are really trying hard to have this out no later than mid-February. Really sorry again for this inconvenience, and will keep you updated with changes now that we're back from vacation

**Comment by deaktomi** [comment]
Thanks for the reply. We keep waiting for the update. I really hope that it will be released in February. We don't want to change our documentation as we want to keep it as simple as possible. That is the requirement of our business. Of course, we'll add a troubleshooting page, but on the main getting started guide we have to commit using long codes.

I hope that you understand and your team can release this soon

**Comment by iscai-msft** [comment]
@deaktomi we definitely understand and are treating this as a high priority. Again really sorry for this delay on our side. The reason for our back and forth is because of potential security concerns with accidentally exposing secret information in the error response string, so we want to make sure we're able to properly address and balance the concern of not having enough information in error response strings and not outputting information that is private.

**Comment by deaktomi** [comment]
Dear @iscai-msft

I'm trying to make work code above but it is not working. To use
```
from azure.core.exceptions import HttpResponseError

try:
    # my code
except HttpResponseError as e:
    print("\n".join([str(e), e.response.text()]))
```
I try to
```
pip install azure-core
```
I see the output:
```
Requirement already satisfied: azure-core in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (1.21.1)
Requirement already satisfied: requests>=2.18.4 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from azure-core) (2.25.1)
Requirement already satisfied: six>=1.11.0 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from azure-core) (1.16.0)
Requirement already satisfied: chardet<5,>=3.0.2 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from requests>=2.18.4->azure-core) (4.0.0)
Requirement already satisfied: certifi>=2017.4.17 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from requests>=2.18.4->azure-core) (2021.5.30)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from requests>=2.18.4->azure-core) (1.26.5)
Requirement already satisfied: idna<3,>=2.5 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from requests>=2.18.4->azure-core) (2.10)
```
However I still receive error when I run my code that is similar to 
```
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    from azure.core.exceptions import HttpResponseError
ModuleNotFoundError: No module named 'azure.core'
```
According to [this page](https://docs.microsoft.com/en-us/python/api/azure-core/azure.core.exceptions?view=azure-python) it seems like I have installed the correct package. I'm I missing something?

**Comment by iscai-msft** [comment]
your pip installation looks correct, so it looks like azure-core is installed within your python39. Based off of the fact that you don't have access to azure core when running your tests, it seems like you might not be using that exact same environment and python version to run your tests. 

To make sure `azure-core` is installed in the environment you use to run your tests, you can check with `pip show azure-core`. If it doesn't show up, you can pip install it again (or just pip install to being with). 

If this still doesn't work, it might be because you're running your code within a different code editor, and the environment your code editor is using is different that then one where azure core is installed. In that case, make sure the virtual environment for your code editor has azure-core installed on it too

**Comment by deaktomi** [comment]
Thanks for the clarification. I was using PyCharm and the debugger used another _venv_ than my terminal. That caused the issue. Package was downloaded to python 3.9 folder. But my debugger used Python 3.8

I could now use the example code you sent up there but unfortunately the `except` was not catching the error. Seems like another unhandled exception is coming earlier, preventing me to catch the `HttpResponseError`

It is coming from the code that is generated by autorest.
```
        # Construct and send request
        request = self._client.get(url, query_parameters, header_parameters)
        response = self._client.send(request, stream=False, **operation_config)

        if response.status_code not in [200]:
            raise models.ProblemResponseException(self._deserialize, response) # Exception is coming from here

        deserialized = None

        if response.status_code == 200:
            deserialized = self._deserialize('StatusResponse', response)
```
The debugger stops on the line with the comment and I can hover my mouse to `response` and I see `content` property where all the API reply is readable:
```
b'{"code":409, "error":{"global":"My error message from API"]}}'
```

Can this code be wrong? Am I missing something?

**Comment by iscai-msft** [comment]
Ok sweet, I think by your update it's working now. Still adding my two cents, it looks like the response has a status code of `200`, even though it's an error response, otherwise it wouldn't make it to that deserialization line. As an update for the error PR it's coming along nicely, will release it beginning of February as part of our February sprint. Thank you again for your patience!

**Comment by iscai-msft** [comment]
@deaktomi the PR fixing this issue is merged into our main branch and will be released as part of our February release cycle, the first week of February. I want to thank you again for your patience, and will reping this issue when azure core is released. Thanks!

**Comment by iscai-msft** [comment]
@deaktomi this fix has been released as part of `azure-core` version 1.22.0! You can update your own version with `pip install azure-core -U`, and you should pin the `azure-core` dependency in the setup.py of your generated package to `"azure-core<2.0.0,>=1.22.0"` so your users get this azure core version as well.

Thank you again so much for your patience! Please try it out and let us know your thoughts!

**Comment by deaktomi** [comment]
@iscai-msft 
Thank you for the update I did the testing and something is still missing on my side.

I run the code with try-catch logic. It was working with nice error message even before the update.
```
try:
    my_client.my_method() # I call the generated library
except MyException as e:
    raise Exception([str(e), json.dumps(e.error.error)])
```
**Output:**
```
C:\Users\MyUser\myproject>python app.py
Traceback (most recent call last):
  File "C:\Users\MyUser\myproject\app.py", line 16, in <module>
    result = client.my_method(
  File "C:\Users\MyUser\AppData\Local\Programs\Python\Python39\lib\site-packages\my_company\my_generated_client\my_client.py", line 188, in my_method
    raise models.MyException(self._deserialize, response)
my_company.my_generated_client.models.my_exception_py3.MyException: Operation returned an invalid status code 'Bad Request'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\MyUser\Dmyproject\app.py", line 42, in <module>
    raise Exception([str(e), json.dumps(e.error.error)])
Exception: ["Operation returned an invalid status code 'Bad Request'", '{"myCustomErrors": ["The error message coming from my api"]}']
```
See the last line with exact error message coming from the API


**Now I update azure-core and run the code without try-catch logic, but the error message is not so nice**
```
C:\Users\MyUser\myproject>pip install azure-core -U
Requirement already satisfied: azure-core in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (1.22.0)
Requirement already satisfied: six>=1.11.0 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from azure-core) (1.16.0)
Requirement already satisfied: requests>=2.18.4 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from azure-core) (2.25.1)
Requirement already satisfied: chardet<5,>=3.0.2 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from requests>=2.18.4->azure-core) (4.0.0)
Requirement already satisfied: idna<3,>=2.5 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from requests>=2.18.4->azure-core) (2.10)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from requests>=2.18.4->azure-core) (1.26.5)
Requirement already satisfied: certifi>=2017.4.17 in c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages (from requests>=2.18.4->azure-core) (2021.5.30)
WARNING: You are using pip version 21.3.1; however, version 22.0.3 is available.
You should consider upgrading via the 'c:\users\myuser\appdata\local\programs\python\python39\python.exe -m pip install --upgrade pip' command.

C:\Users\MyUser\myproject>python app.py
Traceback (most recent call last):
  File "C:\Users\MyUser\myproject\app.py", line 15, in <module>
    result = client.my_method(
  File "C:\Users\MyUser\AppData\Local\Programs\Python\Python39\lib\site-packages\my_company\my_generated_client\my_client.py", line 188, in my_method
    raise models.MyException(self._deserialize, response)
my_company.my_client.models.my_exception_py3.MyException: Operation returned an invalid status code 'Bad Request'
```
As you can see the error message is missing.

I'm pretty sure that this time I have no issue with the python environments.

Do I miss something here?

**Comment by deaktomi** [comment]
I just used cmd.exe on Windows and navigated to my project folder. Please see the results of the command below

```
C:\Users\MyUser\MyProject>dir
 Volume in drive C has no label.
 Volume Serial Number is B281-78B4

 Directory of C:\Users\MyUser\MyProject

2022. 02. 07.  15:31    <DIR>          .
2022. 02. 07.  15:31    <DIR>          ..
2022. 02. 07.  15:30             1 552 app.py
2021. 07. 21.  13:01                24 requirements.txt
               2 File(s)          1 576 bytes
               2 Dir(s)  42 884 505 600 bytes free

C:\Users\MyUser\MyProject>pip show azure-core
Name: azure-core
Version: 1.22.0
Summary: Microsoft Azure Core Library for Python
Home-page: https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/core/azure-core
Author: Microsoft Corporation
Author-email: azpysdkhelp@microsoft.com
License: MIT License
Location: c:\users\myuser\appdata\local\programs\python\python39\lib\site-packages
Requires: requests, six
Required-by:

C:\Users\MyUser\MyProject>python app.py
Traceback (most recent call last):
  File "C:\Users\MyUser\MyProject\app.py", line 15, in <module>
    result = client.my_method(
  File "C:\Users\MyUser\AppData\Local\Programs\Python\Python39\lib\site-packages\my_company\my_client.py", line 188, in my_method
    raise models.MyException(self._deserialize, response)
my_company.my_client.models.my_exception_py3.MyException: Operation returned an invalid status code 'Bad Request'
```

Content of requirements.txt
```
my_company.my_library
```

Contents of app.py
```
from my_company.my_client import MyClient

client = MyClient('http://localhost:9000')

result = client.my_method(
    parameter='wrong_parameter'
)
```

I believe that there can be no different environments here when I call `pip show azure-core` and when I call `python app.py` but I still miss the detailed logs

**Comment by deaktomi** [comment]
**UPDATE**

I tried the very same thing in a clean environment as well.

First I run `pip install my_company.my_client`. The script runs successfully.

Then I run `pip show azure-core`. The script tells me it is not installed. However, my code is working but there are no detailed error message.

Then I run `pip install azure-core`. Seems like 1.22 is installed then. But my code has the same results (no detailed error message)

In my package's `setup.py` there is `install_requires=['msrest'],`. I guess later I should add "azure-core<2.0.0,>=1.22.0" as well but if I install it manually it is not a problem I guess

**Comment by iscai-msft** [comment]
@deaktomi I think I see the issue, you appear to be raising your generated error model as an error itself, see this line: `raise models.MyException(self._deserialize, response)`. If you look at the generated code, the error _model_ just gets passed to the actual error class `HttpResponseError` ([here's](https://github.com/Azure/autorest.python/blob/autorestv3/test/vanilla/legacy/Expected/AcceptanceTests/AdditionalProperties/additionalproperties/operations/_pets_operations.py#L244) a test example). Can you try removing your handwritten code `raise models.MyException(self._deserialize, response)`, and just let the generated code naturally raise? I think if you're using `HttpResponseError`, you should see the error body in the str

**Comment by deaktomi** [comment]
My code was generated with this command on the 25th of January. Please note that I used the `--legacy` argument.
```
autorest --verbose --input-file=./swagger.json --python --output-folder=./output --namespace=my_company.my_client --override-client-name=MyClient  --package-name=my_company.my_client --legacy
```
In the generated code I see what you are telling:
```
if response.status_code not in [200]:
    raise models.MyException(self._deserialize, response)
```
This is generated automatically, I have never touch the `my_client.py` file. In my swagger file, a default response is defined for all endpoints with `MyException` model.
```
"responses": {
  "200": {
	"description": "OK.",
	"schema": {
	  "type": "array",
	  "items": {
		"$ref": "#/definitions/MyResult"
	  }
	}
  },
  "default": {
	"description": "Error",
	"schema": {
	  "$ref": "#/definitions/MyException"
	}
  }
}
```

After this, I've tried to regenerate the client without `--legacy` argument
```
AutoRest code generation utility [cli version: 3.0.6187; node: v14.15.4, max-memory: 8192 gb]
```
The generated `my_client.py` file didn't change at all (some other files did change), so the above lines stayed the same
```
if response.status_code not in [200]:
    raise models.MyException(self._deserialize, response)
```
I also tried to update Autorest by calling `npm install -g autorest@latest`. And regenerated the package without `--legacy` flag
```
AutoRest code generation utility [cli version: 3.5.1; node: v14.15.4]
```
In this case, the generation took a longer time, but the `my_client.py` file is still unchanged and the `raise models.MyException(self._deserialize, response)` lines are still in the code.

Does this mean that the whole thing is caused because I define a default error model for my endpoints in the swagger file? Do you have any suggestions? Do I miss something? Should I use the original example then, with the try-catch logic?

**Comment by deaktomi** [comment]
**UPDATE**
I have tried to add manually the lines that you highlight in the example. I checked the generated `my_client.py` file. Here are the imports:
```
from msrest.service_client import SDKClient
from msrest import Configuration, Serializer, Deserializer
from .version import VERSION
from msrest.pipeline import ClientRawResponse
from . import models
```

In you example `map_error` and `HttpResponseError` are imported from `azure.core.exceptions`. There is no sign of it at all in my code. Why do you think it is happening?

**Comment by iscai-msft** [comment]
No problem at all, so just to confirm everything works now? If so, great!

Want to thank you again for all of your patience (and thank you for your patience with me thanking you all the time 😄). Your issue really helped us improve the overall experience of everyone using our client libraries!

And yep, you definitely diagnosed it right, I didn't realize but you were definitely using the old autorest generation, so very glad that upgrading fixed your issue. Thank you so much again!

**Comment by deaktomi** [comment]
Yes, I can confirm that it works now. I think you can close the issue.

Thanks for your patience as well.
